Prose/ProseLib/Sources/AuthenticationFeature/AuthenticationReducer.swift

```diff
@@ -62,30 +91,77 @@ public let authenticationReducer = Reducer<
     AuthenticationState,
     AuthenticationAction,
     AuthenticationEnvironment
-> { state, action, _ in
+> { state, action, environment in
+    struct CancelId: Hashable {}
+
     switch action {
+    case .alertDismissed:
+        state.alert = nil
+
     case .loginButtonTapped:
         print("Log in button tapped")
-        if state.isFormValid {
-            return Effect(value: .loginResult(.success(state.jid)))
-        } else {
-            return Effect(value: .loginResult(.failure(.badCredentials)))
+        return Effect(value: .logIn)
+
+    case .logIn:
+        state.focusedField = nil
+        state.isLoading = true
+
+        let isFormValid = state.isFormValid
+        let jid = state.jid
+        return Effect.task {
+            if isFormValid, !jid.starts(with: "error") {
+                return .loginResult(.success(jid))
+            } else {
+                return .loginResult(.failure(.badCredentials))
+            }
         }
+        .delay(for: .seconds(1), scheduler: environment.mainQueue)
+        .eraseToEffect()
+        .cancellable(id: CancelId())
 //        return environment.login(state.jid, state.password, .proseAppMacOs)
 //            .map(AuthenticationAction.loginResult)
 
+    case let .showPopoverTapped(popover):
+        state.focusedField = nil
+        state.popover = popover
+
+    case .submitTapped(.address):
+        if state.isAddressValid {
+            state.focusedField = .password
+        } else {
+            state.alert = AlertState(
```

Comment:
I would deem pressing return to have the same effect as clicking the login button. When the login button is disabled nothing happens, otherwise the login routine runs. This would save quite a bit of code since the tracking of the focus field could be removed and we'd save some strings that need to be translated.

Reply:
I would push error messages and validation as much into the CoreLib as possible.

IMHO we don't need _any_ validation on the Swift side in the authentication form. Forms that validate while I'm typing are annoying, so that would leave us with a validation on submit. This can be done in the core library which then can throw an error if needed.

Reply:
I was still thiking about a validation in the Core library, but I get your point.

I will remove this validation.

Reply:
I think that, while it's the role of the app to validate that fields are non-empty, it's the role of the core library to do the rest of the validation work (JID validity, and all other server-backed errors), returned as an ENUM.

On the press return to go to next field, on macOS we could do it as such:

1. Press <return> in the JID field, the app focuses on the next field (password)
2. Press <return> in the password field, the app submits the form (if the form is 100% filled).